File: StringToInteger.java
/*
 * 原题链接：https://leetcode.com/problems/string-to-integer-atoi/
 * 参考链接：https://leetcode.com/discuss/32813/java-solution-with-4-steps-explanations
 * 完成日期：2015.08.20
 * 主要知识点：无
 */

public class StringToInteger {
    public static void main(String[] args) {
        /* 测试 */
    }
}

class Solution {
    private int index = 0;
    private int sign = 1;
    private int total = 0;

    public int myAtoi(String str) {
        /* 首先判断字符串是否有效，null或长度为0直接返回 */
        if(str == null || str.length() == 0) {
            return 0;
        }

        /* 消除前缀空格 */
        while (str.charAt(index) == ' ' && index < str.length()) {
            index++;
        }

        /* 处理符号 */
        if(str.charAt(index) == '+' || str.charAt(index) == '-'){
            sign = str.charAt(index) == '+' ? 1 : -1;
            index++;
        }

        /* 开始转换 */
        while(index < str.length()){
            int digit = str.charAt(index) - '0';
            if (digit < 0 || digit > 9) {
                break;
            }

            /*
             * 判断是否溢出，
             * 这一步很重要，
             * 尤其是后面的取余数判断，
             * 因为单纯的除以10不足以判断是否下一步的相加操作就是溢出；
             * 这个地方还有一种比较讨巧的方法是用double或者long来做，
             * 这样就可以直接和Integer.MAX_VALUE来比较了，
             * 不过总觉得这样做不太好
             */
            if(Integer.MAX_VALUE / 10 < total || Integer.MAX_VALUE / 10 == total && Integer.MAX_VALUE % 10 < digit) {
                return sign == 1 ? Integer.MAX_VALUE : Integer.MIN_VALUE;
            }

            total = 10 * total + digit;
            index++;
        }

        return total * sign;
    }
}

File: kotlin/src/StringToInteger.kt
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2018 Matthew Lee
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

/*
 * https://leetcode.com/problems/string-to-integer-atoi/
 *
 * Implement atoi which converts a string to an integer.
 *
 * The function first discards as many whitespace characters as necessary until the first non-whitespace character is found.
 * Then, starting from this character, takes an optional initial plus or minus sign followed by as many numerical digits as possible,
 * and interprets them as a numerical value.
 *
 * The string can contain additional characters after those that form the integral number,
 * which are ignored and have no effect on the behavior of this function.
 *
 * If the first sequence of non-whitespace characters in str is not a valid integral number,
 * or if no such sequence exists because either str is empty or it contains only whitespace characters,
 * no conversion is performed.
 *
 * If no valid conversion could be performed, a zero value is returned.
 *
 * Note:
 * 1. Only the space character ' ' is considered as whitespace character.
 * 2. Assume we are dealing with an environment which could only store integers within the 32-bit signed integer range: [−2^31,  2&31 − 1].
 *    If the numerical value is out of the range of representable values, INT_MAX (2^31 − 1) or INT_MIN (−2^31) is returned.
 *
 * Example 1:
 * Input: "42"
 * Output: 42
 *
 * Example 2:
 * Input: "   -42"
 * Output: -42
 * Explanation: The first non-whitespace character is '-', which is the minus sign.
 *              Then take as many numerical digits as possible, which gets 42.
 *
 * Example 3:
 * Input: "4193 with words"
 * Output: 4193
 * Explanation: Conversion stops at digit '3' as the next character is not a numerical digit.
 *
 * Example 4:
 * Input: "words and 987"
 * Output: 0
 * Explanation: The first non-whitespace character is 'w', which is not a numerical digit or a +/- sign.
 *              Therefore no valid conversion could be performed.
 *
 * Example 5:
 * Input: "-91283472332"
 * Output: -2147483648
 * Explanation: The number "-91283472332" is out of the range of a 32-bit signed integer.
 *              Therefore INT_MIN (−2^31) is returned.
 */
class StringToInteger {
    fun myAtoi(str: String?): Int {
        if (str == null || str.isEmpty() || str.isBlank()) {
            return 0
        }

        var index = 0
        var total = 0L
        var sign = 1

        while (str[index] == ' ' && index < str.length) {
            index++
        }

        if (str[index] == '+' || str[index] == '-') {
            sign = if (str[index] == '+') { 1 } else { -1 }
            index++
        }

        while (index < str.length) {
            val digit = str[index] - '0'
            if (digit < 0 || digit > 9) {
                break
            }

            if (total > Int.MAX_VALUE) {
                break
            }

            total = 10 * total + digit
            index++
        }

        return if (total > Int.MAX_VALUE) {
            if (sign == 1) { Int.MAX_VALUE } else { Int.MIN_VALUE }
        } else {
            total.toInt() * sign
        }
    }
}
